### PR TITLE
Feature/Xline client watch

### PR DESCRIPTION
**Pull request by bsbds**
Implement watch client in xline-client

Please briefly answer these questions:

* what problem are you trying to solve? (or if there's no problem, what's the motivation for this change?)

* what changes does this pull request make?

* are there any non-obvious implications of these changes? (does it break compatibility with previous versions, etc)

### CODE REVIEW

File: xline-client/src/clients/watch.rs [hunk 1]
```diff
@@ -26,4 +29,267 @@ impl WatchClient {
...
+    ///
+    /// # Panics
+    ///
+    /// If the RPC server returns the wrong value
+    #[inline]
+    pub async fn watch(
+        &mut self,
+        request: WatchRequest,
+    ) -> Result<(Watcher, tonic::Streaming<WatchResponse>)> {
+        let (mut request_sender, request_receiver) = channel::<xlineapi::WatchRequest>(100);
```
Comment (Phoenix500526) [inline xline-client/src/clients/watch.rs:50]: Don't use a magic number. And I think 128 is better.

File: xline-client/src/clients/watch.rs [hunk 2]
```diff
@@ -26,4 +29,267 @@ impl WatchClient {
...
+
+    /// Watches for events happening or that have happened.
+    ///
+    /// # Errors
+    ///
+    /// If sender fails to send to channel
+    #[inline]
+    pub fn watch(&mut self, request: WatchRequest) -> Result<()> {
+        let request = xlineapi::WatchRequest {
+            request_union: Some(RequestUnion::CreateRequest(request.into())),
```
Comment (Phoenix500526) [inline xline-client/src/clients/watch.rs:109]: What are the differences between `WatchClient::watch` and `Watcher::watch`? Both of them send the same `CreateRequest` via `self.sender`. Shall we create twice?

File: xline-client/tests/watch.rs [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    watcher.cancel()?;
+
+    let resp = stream.message().await?.unwrap();
+    assert_eq!(resp.watch_id, watcher.watch_id());
+    assert!(resp.canceled);
+
+    Ok(())
+}
+
+async fn get_cluster_client() -> Result<(Cluster, Client)> {
```
Comment (Phoenix500526) [inline xline-client/tests/watch.rs:38]: The `get_cluster_client` is only used in `watch.rs`. If you don't use the cluster returned by the `get_cluster_client`, then don't return it.

File: xline-client/src/clients/watch.rs (hunk 2 above)
Comment (bsbds) [reply xline-client/src/clients/watch.rs:109]: The Watcher is used to add new watch request or cancel existing request interactively, while WatchClient::watch can only watch on a const range of keys.

File: xline-client/tests/watch.rs (hunk 3 above)
Comment (bsbds) [reply xline-client/tests/watch.rs:38]: The cluster will stop itself when dropped, so we must move it to the test funcitons.

File: xline-client/src/clients/watch.rs (hunk 1 above)
Comment (bsbds) [reply xline-client/src/clients/watch.rs]: done

File: xline-client/src/clients/watch.rs [hunk 4]
```diff
@@ -26,4 +31,268 @@ impl WatchClient {
...
+    #[inline]
+    pub fn cancel_by_id(&mut self, watch_id: i64) -> Result<()> {
+        let request = xlineapi::WatchRequest {
+            request_union: Some(RequestUnion::CancelRequest(WatchCancelRequest { watch_id })),
+        };
+
+        self.sender
+            .try_send(request)
+            .map_err(|e| ClientError::WatchError(e.to_string()))
+    }
```
Comment (themanforfree) [inline xline-client/src/clients/watch.rs:152]: Can `Watcher` cancel other `Watcher`s ?

File: xline-client/src/clients/watch.rs [hunk 5]
```diff
@@ -26,4 +31,268 @@ impl WatchClient {
...
+                key: key.into(),
+                ..Default::default()
+            },
+        }
+    }
+
+    /// Set `key` and `range_end` when with prefix
+    #[inline]
+    #[must_use]
+    pub fn with_prefix(mut self) -> Self {
```
Comment (themanforfree) [inline xline-client/src/clients/watch.rs:195]: Same as auth client

File: xline-client/tests/watch.rs (hunk 3 above)
Comment (themanforfree) [reply xline-client/tests/watch.rs:38]: I suggest use this function in all clients tests, no need to write this function for every test

File: xline-client/src/clients/watch.rs (hunk 4 above)
Comment (bsbds) [reply xline-client/src/clients/watch.rs:152]: I think a `Watcher` is only responsible for the one watch request stream created by `WatchClient::watch`. So If two watcher is created they will not interfere with each other.

File: xline-client/src/clients/watch.rs (hunk 5 above)
Comment (bsbds) [reply xline-client/src/clients/watch.rs]: done

File: xline-client/tests/watch.rs (hunk 3 above)
Comment (bsbds) [reply xline-client/tests/watch.rs]: done

**Review by Phoenix500526 — APPROVED** [review]

**Review by themanforfree — APPROVED** [review]

**Pull request merged**